Jint/Engine.cs

```diff
@@ -452,30 +412,10 @@ public JsValue GetCompletionValue()
 
         internal void RunBeforeExecuteStatementChecks(Statement statement)
         {
-            if (_maxStatements > 0 && _statementsCount++ > _maxStatements)
-            {
-                ExceptionHelper.ThrowStatementsCountOverflowException();
-            }
-
-            if (_timeoutTicks > 0 && _timeoutTicks < DateTime.UtcNow.Ticks)
-            {
-                ExceptionHelper.ThrowTimeoutException();
-            }
-
-            if (_memoryLimit > 0)
+            // Avoid allocating the enumerator because we run this loop very often.
+            for (var i = 0; i < _constraints.Count; i++)
```

Comment:
Another idea that came to my mind to copy the constrains engine's array, then foreach is as fast as for loop and loop against array is always faster than against list. The options might even keep track of the constraints changing and invalidate the backing array and generate on request (Constrains getter) - options would be reusable and efficient. Might be future improvements.

Reply:
I am not sure if it makes a big difference...the list is an array under the hood and I guess the compiler will inline some of the calls to the list indexer anyway.